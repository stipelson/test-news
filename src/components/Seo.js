/**
 * SEO component that queries for data with
 *  Gatsby's useStaticQuery React hook
 *
 * See: https://www.gatsbyjs.org/docs/use-static-query/
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Helmet } from 'react-helmet';
import { useStaticQuery, graphql } from 'gatsby';

export const PureSEO = ({ description, lang, meta, title, data }) => {
  const metaDescription =
    description || (data && data.site)
      ? data.site.siteMetadata.description
      : 'Site description';

  const dataTitle =
    data && data.site ? data.site.siteMetadata.title : 'Site title';

  const dataAuthor =
    data && data.site ? data.site.siteMetadata.author : 'Site title';

  return (
    <Helmet
      htmlAttributes={{
        lang,
      }}
      title={title}
      titleTemplate={`%s | ${dataTitle}`}
      meta={[
        {
          name: 'description',
          content: metaDescription,
        },
        {
          property: 'og:title',
          content: title,
        },
        {
          property: 'og:description',
          content: metaDescription,
        },
        {
          property: 'og:type',
          content: 'website',
        },
        {
          name: 'twitter:card',
          content: 'summary',
        },
        {
          name: 'twitter:creator',
          content: dataAuthor,
        },
        {
          name: 'twitter:title',
          content: title,
        },
        {
          name: 'twitter:description',
          content: metaDescription,
        },
      ].concat(meta)}
    />
  );
};

PureSEO.defaultProps = {
  lang: 'en',
  meta: [],
  description: '',
};

PureSEO.propTypes = {
  description: PropTypes.string,
  lang: PropTypes.string,
  meta: PropTypes.arrayOf(PropTypes.object),
  title: PropTypes.string.isRequired,
  data: PropTypes.object,
};

export const SEO = (props) => {
  const data = useStaticQuery(graphql`
    query {
      site {
        siteMetadata {
          title
        }
      }
    }
  `);
  return <PureSEO {...props} data={data}></PureSEO>;
};

export default SEO;
