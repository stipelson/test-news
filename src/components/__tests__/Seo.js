import React from 'react';
import { create, act } from 'react-test-renderer';
import { PureSEO as SEO } from '../Seo';
import { Helmet } from 'react-helmet';

describe('SEO render', () => {
  let component;
  let data;
  let meta;

  beforeAll(() => {
    data = {
      site: {
        siteMetadata: {
          title: 'Site New Title',
          description: 'Site New description',
          author: 'Site New Author',
        },
      },
    };

    meta = [
      {
        name: 'description',
        content: data.site.siteMetadata.description,
      },
      {
        property: 'og:title',
        content: data.site.siteMetadata.title,
      },
      {
        property: 'og:description',
        content: data.site.siteMetadata.description,
      },
      {
        property: 'og:type',
        content: 'website',
      },
      {
        name: 'twitter:card',
        content: 'summary',
      },
      {
        name: 'twitter:creator',
        content: data.site.siteMetadata.author,
      },
      {
        name: 'twitter:title',
        content: data.site.siteMetadata.title,
      },
      {
        name: 'twitter:description',
        content: data.site.siteMetadata.description,
      },
    ];
    component = create(<SEO title="Site title" />);
  });

  it('renders correctly', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('Verify helmet default data', () => {
    let instance = component.root;
    const helmet = instance.findByType(Helmet);

    expect(helmet.props.title).toEqual('Site title');

    expect(helmet.props.titleTemplate).toEqual('%s | Site title');

    const { lang } = helmet.props.htmlAttributes;
    expect(lang).toEqual('en');
  });

  it('Verify helmet exist with custom params', () => {
    act(() => {
      component.update(<SEO data={data} title="Site title" />);
    });
    let instance = component.root;
    const helmet = instance.findByType(Helmet);

    expect(helmet.props.title).toEqual('Site title');
  });

  it('Verify helmet custom props', () => {
    act(() => {
      component.update(<SEO data={data} title="Test title" lang="TEST" />);
    });
    let instance = component.root;
    const helmet = instance.findByType(Helmet);

    expect(helmet.props.title).toEqual('Test title');

    expect(helmet.props.titleTemplate).toEqual('%s | Site New Title');

    const { lang } = helmet.props.htmlAttributes;
    expect(lang).toEqual('TEST');
  });

  it('Helmet verify meta', () => {
    act(() => {
      component.update(
        <SEO data={data} title={data.site.siteMetadata.title} />
      );
    });
    let instance = component.root;
    const helmet = instance.findByType(Helmet);

    expect(helmet.props.meta).toEqual(meta);
  });

  it('Helmet verify meta concat', () => {
    let moreMeta = [{ property: 'OtherMeta', content: 'test meta' }];
    act(() => {
      component.update(
        <SEO data={data} title={data.site.siteMetadata.title} meta={moreMeta} />
      );
    });
    let instance = component.root;
    const helmet = instance.findByType(Helmet);

    expect(helmet.props.meta).toEqual(meta.concat(moreMeta));
  });
});
