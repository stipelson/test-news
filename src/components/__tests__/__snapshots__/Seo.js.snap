// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SEO render renders correctly 1`] = `null`;
